### PR TITLE
♻️ Refactor math utils in extensions

### LINKED ISSUE #21104: Refactor various math utility functions
There are several simple math utility functions that could be consolidated to one place. For example, code to calculate distance like

`Math.sqrt((dx * dx) + (dy * dy));`

shows up in `amp-pan-zoom`, `amp-image-viewer`, `amp-video-docking`, `amp-image-lightbox` and soon `amp-lightbox-gallery`.

Another example is some places that don't use the existing `clamp` utility function: `amp-image-viewer`, `amp-image-slider`, `amp-image-lightbox` and a few more.

### PR DESCRIPTION
**Pull request by rcebulko**
Pull common math functionality into helper utils.
- Adds helpers to bound values with extent, compute vector magnitude, and compute distance
- Removes duplicate implementations of math functions
- Updates `amp-image-slider`, `amp-image-lightbox`, `amp-lightbox-gallery`, `amp-image-viewer`, and `amp-video-docking` to use helpers where appropriate.

> Note: Will be easiest to review commit-by-commit

Fixes #21104

### CODE REVIEW

**Comment by rcebulko** [comment]
I tried running `gulp test --unit --files=<this js file>` but quickly realized that's (obviously) not where the tests are defined. Is there a canonical mapping of src to specs?

**Comment by cathyxz** [comment]
@rcebulko not exactly (though that would be nice). The `--files` param is used to refer to the test file that you want to run, for example in this case, the relevant test file is `test/unit/utils/test-math.js`, so you could run `gulp test --unit --files=test/unit/utils/test-math.js`. The relevant test file is usually named `test-[name-of-library-or-extension].js`, which makes it a bit easier to find (hopefully).

**Comment by rcebulko** [comment]
Worth noting: if I completely bork these utility functions, the only extension that fails specs is amp-image-lightbox. The rest pass normally.

@cathyxz Thanks for filling me in! I found em and added relevant specs.

**Comment by rsimha** [comment]
Nice! I'll let @sparhami review the actual library changes.

Now that you have moved all `Math.*` functions to `math.js`, I think it's worth adding a presubmit rule to forbid the use of the term in normal source files. This is a good place:

https://github.com/ampproject/amphtml/blob/613b5e257e1c7263f5be636c073c903414e9c2d3/build-system/tasks/presubmit-checks.js#L45-L46

**Comment by rcebulko** [comment]
> Nice! I'll let @sparhami review the actual library changes.
> 
> Now that you have moved all `Math.*` functions to `math.js`, I think it's worth adding a presubmit rule to forbid the use of the term in normal source files. This is a good place:
> 
> [amphtml/build-system/tasks/presubmit-checks.js](https://github.com/ampproject/amphtml/blob/613b5e257e1c7263f5be636c073c903414e9c2d3/build-system/tasks/presubmit-checks.js#L45-L46)
> 
> Lines 45 to 46 in [613b5e2](/ampproject/amphtml/commit/613b5e257e1c7263f5be636c073c903414e9c2d3)
> 
>  // Terms that must not appear in our source files. 
>  const forbiddenTerms = {

Do you mean forbidding use of /Math\.(min|max|pow|sqrt)/? Cause all of them could have legitimate uses on their own, it seems like we'd only want to forbid particular compositions of them, which may not be a very plausible check to implement in practice (ex. some recent code assigned to `dx` and `dy` and squared them, rather than following a particular format/pattern). It might be possible to write up a function to match, but it seems like it would be an imperfect solution that would be better solved by an awareness by reviewers of existing utility functions (which applies across the board for common methods).

**Comment by rsimha** [comment]
Re: presubmit checks, fair enough.

**Comment by rcebulko** [comment]
I'm not sure how to make sense of the Percy diff/warning.

**Comment by rcebulko** [comment]
How flaky/noisy are the Travis CI checks usually, specifically the _Visual Diff_, _Remote (Sauce Labs) Tests_, and _End to End_ tests? I'm seeing very odd errors that seem pretty unrelated to my changes.

**Comment by danielrozenberg** [comment]
Pretty flaky :( we're actively working on fixing this though. I'll restart the failing Visual Diffs shard (the other two shards are non-fatal, and are allowed to fail on Travis without failing the entire PR)

**Comment by rcebulko** [comment]
> Pretty flaky :( we're actively working on fixing this though. I'll restart the failing Visual Diffs shard (the other two shards are non-fatal, and are allowed to fail on Travis without failing the entire PR)

Is there any particular thread or channel used to track common/currently-failing checks, or should I trust my intuition + reviewers?

File: src/utils/math.js [hunk 1]
```diff
@@ -95,3 +95,37 @@ export function mod(a, b) {
...
+/**
+ * Returns value bound to min and max values +/- extent.
+ * @param {number} val the value to bound.
+ * @param {number} min the lower bound.
+ * @param {number} max the upper bound
+ * @param {number} extent the allowed extent beyond the bounds.
+ * @return {number} the bounded value.
+ */
+export function boundValue(val, min, max, extent) {
+  return clamp(val, min - extent, max + extent);
```
Comment (sparhami) [inline src/utils/math.js]: Probably worth adding a dev assert that min <= max.

Comment (rcebulko) [reply src/utils/math.js]: If that's the case, it seems to make the most sense to put that in `clamp` rather than here

Comment (rcebulko) [reply src/utils/math.js]: Added to both

File: test/unit/utils/test-math.js [hunk 2]
```diff
@@ -124,9 +124,15 @@ describes.sandboxed('utils/math', {}, () => {
     });
 
     it('should fail if the minimum is greater than the maximum', () => {
-      expect(() => clamp(1, 1, 0)).to.throw();
-      expect(() => clamp(0.3, 0.5, 0.1)).to.throw();
-      expect(() => clamp(0, 10, -10)).to.throw();
+      allowConsoleError(() => {
```
Comment (rcebulko) [inline test/unit/utils/test-math.js]: Without this, Travis will fail (even though local may not)

**Comment by rsimha** [comment]
@rcebulko Nice work! Is this good to merge?

File: test/unit/utils/test-math.js [hunk 3]
```diff
@@ -114,5 +121,132 @@ describes.sandboxed('utils/math', {}, () => {
...
+    it('should fail if the minimum is greater than the maximum', () => {
+      allowConsoleError(() => {
+        expect(() => clamp(1, 1, 0)).to.throw();
+      });
+      allowConsoleError(() => {
+        expect(() => clamp(0.3, 0.5, 0.1)).to.throw();
+      });
+      allowConsoleError(() => {
+        expect(() => clamp(0, 10, -10)).to.throw();
+      });
```
Comment (rsimha) [inline test/unit/utils/test-math.js]: nit: This block can be simplified to:
```js
    it('should fail if the minimum is greater than the maximum', () => {
      allowConsoleError(() => {
        expect(() => clamp(1, 1, 0)).to.throw();
        expect(() => clamp(0.3, 0.5, 0.1)).to.throw();
        expect(() => clamp(0, 10, -10)).to.throw();
      });
    });
```

**Pull request merged**